### SPEED DATING EXPERIMENT

In [40]:
import os
import numpy as np
import pandas as pd
import csv
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
import statsmodels.formula.api as smf
from sklearn import linear_model, metrics, cross_validation, feature_selection

pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

%matplotlib inline
plt.style.use('ggplot')



pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

In [41]:
def linear_modeling_with_sklearn(X, y):
    model = linear_model.LinearRegression(fit_intercept = True)
    model.fit(X, y)

    print 'F-statistic (performed for each regressor independently)'
    print '- F-value', feature_selection.f_regression(X, y)[0]
    print '- p-value', feature_selection.f_regression(X, y)[1]
    print 'R^2 =', model.score(X, y)
    print 'Coefficients'
    print '- beta_0 (intercept) =', model.intercept_
    print '- beta_n (n > 0)     =', model.coef_

In [42]:
df = pd.read_csv(os.path.join('..', 'CODE', 'speed-dating-experiment', 'Speed Dating Data.csv'))

In [43]:
df

iid    id  gender  idg  condtn   ...    attr5_3  sinc5_3  intel5_3  \
0       1   1.0       0    1       1   ...        NaN      NaN       NaN   
1       1   1.0       0    1       1   ...        NaN      NaN       NaN   
2       1   1.0       0    1       1   ...        NaN      NaN       NaN   
3       1   1.0       0    1       1   ...        NaN      NaN       NaN   
4       1   1.0       0    1       1   ...        NaN      NaN       NaN   
...   ...   ...     ...  ...     ...   ...        ...      ...       ...   
8373  552  22.0       1   44       2   ...        9.0      5.0       9.0   
8374  552  22.0       1   44       2   ...        9.0      5.0       9.0   
8375  552  22.0       1   44       2   ...        9.0      5.0       9.0   
8376  552  22.0       1   44       2   ...        9.0      5.0       9.0   
8377  552   NaN       1   44       2   ...        9.0      5.0       9.0   

      fun5_3  amb5_3  
0        NaN     NaN  
1        NaN     NaN  
2        NaN     NaN  
3        NaN     NaN  
4        NaN     NaN  
...      ...     ...  
8373     5.0     6.0  
8374     5.0     6.0  
8375     5.0     6.0  
8376     5.0     6.0  
8377     5.0     6.0  

[8378 rows x 195 columns]

In [44]:
df.columns

Index([u'iid', u'id', u'gender', u'idg', u'condtn', u'wave', u'round',
       u'position', u'positin1', u'order',
       ...
       u'attr3_3', u'sinc3_3', u'intel3_3', u'fun3_3', u'amb3_3', u'attr5_3',
       u'sinc5_3', u'intel5_3', u'fun5_3', u'amb5_3'],
      dtype='object', length=195)

In [45]:
sorted(df.wave.unique())

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21]

In [46]:
df.wave.value_counts()

21    968
11    882
9     800
14    720
15    684
     ... 
10    162
16     96
20     84
18     72
6      50
Name: wave, dtype: int64

In [47]:
df.corr()

iid        id    gender       idg    condtn    ...     \
iid       1.000000  0.117109  0.051937  0.120362  0.220413    ...      
id        0.117109  1.000000 -0.011623  0.998029  0.318768    ...      
gender    0.051937 -0.011623  1.000000  0.033645 -0.002627    ...      
idg       0.120362  0.998029  0.033645  1.000000  0.324838    ...      
condtn    0.220413  0.318768 -0.002627  0.324838  1.000000    ...      
...            ...       ...       ...       ...       ...    ...      
attr5_3   0.092325 -0.018183 -0.103692 -0.020785  0.068552    ...      
sinc5_3  -0.053698 -0.007993 -0.289694 -0.022967 -0.063407    ...      
intel5_3  0.006949 -0.088398  0.076538 -0.086071  0.068622    ...      
fun5_3    0.100164 -0.046558 -0.067432 -0.050316  0.114254    ...      
amb5_3    0.055229 -0.119108  0.055267 -0.118940  0.145598    ...      

           attr5_3   sinc5_3  intel5_3    fun5_3    amb5_3  
iid       0.092325 -0.053698  0.006949  0.100164  0.055229  
id       -0.018183 -0.007993 -0.088398 -0.046558 -0.119108  
gender   -0.103692 -0.289694  0.076538 -0.067432  0.055267  
idg      -0.020785 -0.022967 -0.086071 -0.050316 -0.118940  
condtn    0.068552 -0.063407  0.068622  0.114254  0.145598  
...            ...       ...       ...       ...       ...  
attr5_3   1.000000  0.197948  0.378176  0.406725  0.205368  
sinc5_3   0.197948  1.000000  0.505440  0.331171  0.268553  
intel5_3  0.378176  0.505440  1.000000  0.270181  0.424972  
fun5_3    0.406725  0.331171  0.270181  1.000000  0.398011  
amb5_3    0.205368  0.268553  0.424972  0.398011  1.000000  

[187 rows x 187 columns]

In [48]:
df['round'].value_counts()

18    1350
20    1160
22     968
21     882
10     880
      ... 
6      162
11     140
5       50
8       48
7       42
Name: round, dtype: int64

In [49]:
sorted(df['round'].unique())

[5, 6, 7, 8, 9, 10, 11, 14, 15, 16, 18, 19, 20, 21, 22]

In [50]:
sorted(df.position.unique())

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22]

# Look at Match (1=yes, 0 = no)

In [51]:
df.corr().match

iid        -0.015806
id         -0.000770
gender     -0.000530
idg        -0.000544
condtn     -0.045960
              ...   
attr5_3     0.023098
sinc5_3    -0.031990
intel5_3   -0.013012
fun5_3      0.019793
amb5_3      0.008534
Name: match, dtype: float64

### Look at Match and Age

In [52]:
model = smf.ols(formula = 'match ~ age ', data = df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  match   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     10.19
Date:                Thu, 23 Jun 2016   Prob (F-statistic):            0.00142
Time:                        17:54:49   Log-Likelihood:                -3527.5
No. Observations:                8283   AIC:                             7059.
Df Residuals:                    8281   BIC:                             7073.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept      0.2605      0.030      8.579      0.000         0.201     0.320
age           -0.0036      0.001     -3.192      0.001        -0.006    -0.001
==============================================================================
Omnibus:                     2399.087   Durbin-Watson:                   1.835
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5068.646
Skew:                           1.807   Prob(JB):                         0.00
Kurtosis:                       4.274   Cond. No.                         199.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Look at Match and Race

race_o:		race of partner

- Black/African American=1
- European/Caucasian-American=2
- Latino/Hispanic American=3
- Asian/Pacific Islander/Asian-American=4
- Native American=5
- Other=6


__imprace__: How important is it to you (on a scale of 1-10) that a person you date be of the same racial/ethnic background?


In [53]:
model = smf.ols(formula = 'match ~ race_o + race + imprace ', data = df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  match   R-squared:                       0.003
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     8.519
Date:                Thu, 23 Jun 2016   Prob (F-statistic):           1.20e-05
Time:                        17:54:49   Log-Likelihood:                -3491.6
No. Observations:                8226   AIC:                             6991.
Df Residuals:                    8222   BIC:                             7019.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept      0.2154      0.015     14.484      0.000         0.186     0.245
race_o        -0.0035      0.003     -1.058      0.290        -0.010     0.003
race          -0.0055      0.003     -1.647      0.100        -0.012     0.001
imprace       -0.0069      0.001     -4.814      0.000        -0.010    -0.004
==============================================================================
Omnibus:                     2378.725   Durbin-Watson:                   1.838
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5020.322
Skew:                           1.804   Prob(JB):                         0.00
Kurtosis:                       4.277   Cond. No.                         22.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

samerace:  	participant and the partner were the same race. 1= yes, 0=no

In [54]:
model = smf.ols(formula = 'match ~ samerace ', data = df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  match   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     1.422
Date:                Thu, 23 Jun 2016   Prob (F-statistic):              0.233
Time:                        17:54:50   Log-Likelihood:                -3578.2
No. Observations:                8378   AIC:                             7160.
Df Residuals:                    8376   BIC:                             7175.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept      0.1608      0.005     30.843      0.000         0.151     0.171
samerace       0.0099      0.008      1.192      0.233        -0.006     0.026
==============================================================================
Omnibus:                     2425.371   Durbin-Watson:                   1.836
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5122.283
Skew:                           1.807   Prob(JB):                         0.00
Kurtosis:                       4.268   Cond. No.                         2.45
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Look at Match and Field

In [55]:
model = smf.ols(formula = 'match ~ field_cd ', data = df).fit()
model.summary()

#NEED TO TURN INTO DUMMY VALUE

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  match   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     5.039
Date:                Thu, 23 Jun 2016   Prob (F-statistic):             0.0248
Time:                        17:54:50   Log-Likelihood:                -3540.2
No. Observations:                8296   AIC:                             7084.
Df Residuals:                    8294   BIC:                             7098.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept      0.1833      0.009     19.828      0.000         0.165     0.201
field_cd      -0.0024      0.001     -2.245      0.025        -0.005    -0.000
==============================================================================
Omnibus:                     2401.139   Durbin-Watson:                   1.836
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5070.221
Skew:                           1.807   Prob(JB):                         0.00
Kurtosis:                       4.269   Cond. No.                         19.6
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

field_cd: 	field coded 

- 1= Law 
- 2= Math
- 3= Social Science, Psychologist 
- 4= Medical Science, Pharmaceuticals, and Bio Tech 
- 5= Engineering  
- 6= English/Creative Writing/ Journalism 
- 7= History/Religion/Philosophy 
- 8= Business/Econ/Finance 
- 9= Education, Academia 
- 10= Biological Sciences/Chemistry/Physics
- 11= Social Work 
- 12= Undergrad/undecided 
- 13=Political Science/International Affairs 
- 14=Film
- 15=Fine Arts/Arts Administration
- 16=Languages
- 17=Architecture
- 18=Other

# Look at Age, Gender, Goal

### Gender:


- Female=0
- Male=1


In [56]:
df.gender.unique()

array([0, 1])

In [57]:
df.gender.value_counts()

1    4194
0    4184
Name: gender, dtype: int64

In [58]:
len(df.gender.dropna())

8378

### Age:

In [59]:
df.age.unique()

array([ 21.,  24.,  25.,  23.,  22.,  26.,  27.,  30.,  28.,  nan,  29.,
        34.,  35.,  32.,  39.,  20.,  19.,  18.,  37.,  33.,  36.,  31.,
        42.,  38.,  55.])

In [60]:
df.age.value_counts()

27.0    1059
23.0     894
26.0     869
24.0     863
25.0     837
        ... 
38.0      19
39.0      18
18.0      10
55.0       6
37.0       5
Name: age, dtype: int64

### Goal:

goal:
What is your primary goal in participating in this event? 
- Seemed like a fun night out=1
- To meet new people=2
- To get a date=3
- Looking for a serious relationship=4
- To say I did it=5
- Other=6


In [61]:
df.goal.value_counts()

1.0    3426
2.0    3012
3.0     631
5.0     510
6.0     419
4.0     301
Name: goal, dtype: int64

In [62]:
df.goal.unique()

array([  2.,   1.,   6.,   3.,   4.,  nan,   5.])

In [63]:
len(df.goal.dropna())

8299

In [64]:
#look at goal distribution 
print np.nan, df.goal.isnull().sum()
for goals in np.sort(df.goal.dropna().unique()):
    print goals, len(df[df.goal == goals])

nan 79
1.0 3426
2.0 3012
3.0 631
4.0 301
5.0 510
6.0 419


In [65]:
df = df[df.goal.isin([1, 2, 3, 4, 5, 6])]

#counts for goal - NAN values --> only picking 1-6

In [66]:
print np.nan, df.goal.isnull().sum()
df.goal.value_counts()

#removed all NAN values

nan 0


1.0    3426
2.0    3012
3.0     631
5.0     510
6.0     419
4.0     301
Name: goal, dtype: int64

In [67]:
df.goal.value_counts()

#checking that NAN values were removed

1.0    3426
2.0    3012
3.0     631
5.0     510
6.0     419
4.0     301
Name: goal, dtype: int64

In [68]:
sorted(df.goal.unique())

#checking again

[1.0, 2.0, 3.0, 4.0, 5.0, 6.0]

#### *Create Dummy Variables*

In [69]:
goal_df = pd.get_dummies(df.goal, prefix = 'Goal')
goal_df

Goal_1.0  Goal_2.0  Goal_3.0  Goal_4.0  Goal_5.0  Goal_6.0
0          0.0       1.0       0.0       0.0       0.0       0.0
1          0.0       1.0       0.0       0.0       0.0       0.0
2          0.0       1.0       0.0       0.0       0.0       0.0
3          0.0       1.0       0.0       0.0       0.0       0.0
4          0.0       1.0       0.0       0.0       0.0       0.0
...        ...       ...       ...       ...       ...       ...
8373       1.0       0.0       0.0       0.0       0.0       0.0
8374       1.0       0.0       0.0       0.0       0.0       0.0
8375       1.0       0.0       0.0       0.0       0.0       0.0
8376       1.0       0.0       0.0       0.0       0.0       0.0
8377       1.0       0.0       0.0       0.0       0.0       0.0

[8299 rows x 6 columns]

In [70]:
#rename:

goal_df.rename(columns={'Goal_1.0': 'goal_1',
                        'Goal_2.0': 'goal_2',
                        'Goal_3.0': 'goal_3',
                        'Goal_4.0': 'goal_4',
                        'Goal_5.0': 'goal_5',
                        'Goal_6.0': 'goal_6'}, inplace = True)

#### *Join dummy set with dataset and remove 'goal' column:*

In [71]:
df = df.join([goal_df])  #joining dummy variables into dataset

df.drop('goal', axis = 1, inplace = True)  #remove the column 'goal'

### Match ~ Goal + Age + Gender * Age + Gender with StatsModels / sklearn  ???? --> DOESNT WORK?

In [72]:
#X = df[ ['age', 'goal_1'] ]
#y = df.match

#linear_modeling_with_sklearn(X, y)

#smf.ols(formula = 'match ~ goal + age + gender + age * gender', data = df).fit().summary()

## Odds / OR of Goal & Gender

female = 0
male = 1


goal:
What is your primary goal in participating in this event? 
- Seemed like a fun night out=1
- To meet new people=2
- To get a date=3
- Looking for a serious relationship=4
- To say I did it=5
- Other=6



#### Goal 1: Seemed like a fun night out

In [73]:
pd.crosstab(df.gender, df.goal_1, margins=True)

goal_1   0.0   1.0   All
gender                  
0       2438  1687  4125
1       2435  1739  4174
All     4873  3426  8299

Odds of being male & choosing to participate b/c it 'seemed like a fun night out' (which i'd argue indicated the feelings that it was NOT LONG TERM/casual): 1739:2435

Odds of being female & choosing to participate b/c it 'seemed like a fun night out' (casual): 1687:2438

__ODDS RATIO__: (1739/2435) / (1687/2438) = 1.03

*odds of a male looking for a casual night out : odds of being a female and looking for a casual night out*

1.03:1

    Men more likely to be participating for a casual night out but not by much

#### Goal 2: To meet new people

In [74]:
pd.crosstab(df.gender, df.goal_2, margins=True)

goal_2   0.0   1.0   All
gender                  
0       2478  1647  4125
1       2809  1365  4174
All     5287  3012  8299

Odds of being male & choosing to participate to meet new people:  1365:2809

Odds of being female & choosing to participate to meet new people: 1647:2478

__ODDS RATIO__: (1365/2809) / (1647:2478) = .73

*odds of a male looking to meet new people : odds of being a female and looking to meet new people*

.73:1

    Men are 27% less likely to look to meet new people than women

#### Goal 3: To get a date

In [75]:
pd.crosstab(df.gender, df.goal_3, margins=True)

goal_3   0.0  1.0   All
gender                 
0       3928  197  4125
1       3740  434  4174
All     7668  631  8299

Odds of looking trying to get a date & being male:  434:3740

Odds of looking trying to get a date & being female:  197:3928

__ODDS RATIO__: (434/3740) / (197/3928) = 2.314

*odds of a male looking for a date : odds of being a female and looking for a date*

2.314:1

    More men looking for a date than women in this study

#### Goal 4: Looking for a serious relationship

In [76]:
pd.crosstab(df.gender, df.goal_4, margins=True)

goal_4   0.0  1.0   All
gender                 
0       3996  129  4125
1       4002  172  4174
All     7998  301  8299

Odds of looking for a serious relationship and being male:  172:4002

Odds of looking for a serious relationship and not being male:  129:3996

ODDS RATIO: (172/4002) / (129/3996) = 1.33 

*odds of a male looking for a serious relationship : odds of being a female and looking for a serious relationship*

1.33:1

    More men looking for serious relationship than women in this study

#### Goal 5: To say I did it

In [77]:
pd.crosstab(df.gender, df.goal_5, margins=True)

goal_5   0.0  1.0   All
gender                 
0       3849  276  4125
1       3940  234  4174
All     7789  510  8299

Odds of being male & choosing to participate to 'say i did it':  234:3940

Odds of being female & choosing to participate to 'say i did it': 276:3849

__ODDS RATIO__: (234/3940) / (276/3849) = .83

*odds of a male participating to 'say he went on a speed dating experiment' : odds of being a female participating to'say she went on a speed dating experiment'*

.83:1

    Men are 17% less likely to participate in the speed dating experiement just to say they have 
    (observation: perhaps speed dating is more geared towards women)

#### Goal 6: Other

In [78]:
pd.crosstab(df.gender, df.goal_6, margins=True)

goal_6   0.0  1.0   All
gender                 
0       3936  189  4125
1       3944  230  4174
All     7880  419  8299

Odds of choosing 'Other' if male: 230:3944

Odds of choosing 'Other' if female: 189:3936

__ODDS RATIO__: (230/3944) / (189/3936) = 1.22

*odds of a male selecting 'other' as reason for participating: odds of a female selecting 'other' as reason for participating*

1.22:1

    Men are 1.22x more likely to select other when deciding reason for participating in study